### DISCUSSION TITLE
Generate a polycrystal structure,  then stop the grain boundary growth and conduct a mechanical stress analysis.

### DISCUSSION THREAD

**Discussion by Yaode-Yin** [idaholab/moose#27733] [Q&A Modules: Phase field]
Hello, everyone!

I'm planning to generate a polycrystal structure followed the example `poly_grain_growth_2D_eldrforce.i`. After generating the polycrystal structure, I intend to stop the grain boundary growth and proceed with additional mechanical stress analyses. Naturally, the Euler angle information (indicating the orientation of the grains) needs to be preserved. And I have some questions needed to address. 
    1. How can MOOSE stop the `GBEvolution` at a desired time, and then conduct a pure mechanical stress analysis?
    2. How can we utilize `BndsCalcIC` to define grain boundary blocks and assign special material parameters to the interfacial region in the mechanical stress analysis?
    3. If there are any alternative solutions, please do let me know.
I have tried some attempts, but don't find proper method to stop the grain boundary growth. I need the euler angle information of the grains in the mechanical analysis. Thank you!

Best wishes! 
Yaode Yin

**Comment by Yaode-Yin** [comment]
Here is an attempt by using `Control system`. But unfortunately, it doesn't work. The grain growth cannot stop. 

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 20
  ny = 20
  nz = 0
  xmax = 1000
  ymax = 1000
  zmax = 0
  elem_type = QUAD4
  uniform_refine = 2
[]

[GlobalParams]
  op_num = 8
  var_name_base = gr
  grain_num = 36
[]

[Variables]
  [./PolycrystalVariables]
  [../]
  [./disp_x]
    order = FIRST
    family = LAGRANGE
  [../]
  [./disp_y]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[UserObjects]
  [./euler_angle_file]
    type = EulerAngleFileReader
    file_name = grn_36_rand_2D.tex
  [../]
  [./voronoi]
    type = PolycrystalVoronoi
    coloring_algorithm = bt
  [../]
  [./grain_tracker]
    type = GrainTrackerElasticity
    threshold = 0.2
    compute_var_to_feature_map = true
    execute_on = 'initial timestep_begin'
    flood_entity_type = ELEMENTAL

    C_ijkl = '1.27e5 0.708e5 0.708e5 1.27e5 0.708e5 1.27e5 0.7355e5 0.7355e5 0.7355e5'
    fill_method = symmetric9
    euler_angle_provider = euler_angle_file
  [../]
[]

[ICs]
  [./PolycrystalICs]
    [./PolycrystalColoringIC]
      polycrystal_ic_uo = voronoi
    [../]
  [../]
[]

[AuxVariables]
  [./bnds]
    order = FIRST
    family = LAGRANGE
  [../]
  [./elastic_strain11]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./elastic_strain22]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./elastic_strain12]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./unique_grains]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./var_indices]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./vonmises_stress]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./C1111]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./euler_angle]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[Kernels]
  [./PolycrystalKernel]
  [../]
  [./PolycrystalElasticDrivingForce]
  [../]
  [./TensorMechanics]
    use_displaced_mesh = true
    displacements = 'disp_x disp_y'
  [../]
[]

[AuxKernels]
  [./BndsCalc]
    type = BndsCalcAux
    variable = bnds
    execute_on = timestep_end
  [../]
  [./elastic_strain11]
    type = RankTwoAux
    variable = elastic_strain11
    rank_two_tensor = elastic_strain
    index_i = 0
    index_j = 0
    execute_on = timestep_end
  [../]
  [./elastic_strain22]
    type = RankTwoAux
    variable = elastic_strain22
    rank_two_tensor = elastic_strain
    index_i = 1
    index_j = 1
    execute_on = timestep_end
  [../]
  [./elastic_strain12]
    type = RankTwoAux
    variable = elastic_strain12
    rank_two_tensor = elastic_strain
    index_i = 0
    index_j = 1
    execute_on = timestep_end
  [../]
  [./unique_grains]
    type = FeatureFloodCountAux
    variable = unique_grains
    execute_on = timestep_end
    flood_counter = grain_tracker
    field_display = UNIQUE_REGION
  [../]
  [./var_indices]
    type = FeatureFloodCountAux
    variable = var_indices
    execute_on = timestep_end
    flood_counter = grain_tracker
    field_display = VARIABLE_COLORING
  [../]
  [./C1111]
    type = RankFourAux
    variable = C1111
    rank_four_tensor = elasticity_tensor
    index_l = 0
    index_j = 0
    index_k = 0
    index_i = 0
    execute_on = timestep_end
  [../]
  [./vonmises_stress]
    type = RankTwoScalarAux
    variable = vonmises_stress
    rank_two_tensor = stress
    scalar_type = VonMisesStress
    execute_on = timestep_end
  [../]
  [./euler_angle]
    type = OutputEulerAngles
    variable = euler_angle
    euler_angle_provider = euler_angle_file
    grain_tracker = grain_tracker
    output_euler_angle = 'phi1'
    execute_on = 'initial timestep_end'
  [../]
[]

[BCs]
  [./Periodic]
    [./All]
      auto_direction = 'x'
      variable = 'gr0 gr1 gr2 gr3 gr4 gr5 gr6 gr7'
    [../]
  [../]
  [./top_displacement]
    type = DirichletBC
    variable = disp_y
    boundary = top
    value = 0
  [../]
  [./x_anchor]
    type = DirichletBC
    variable = disp_x
    boundary = 'left right'
    value = 0.0
  [../]
  [./y_anchor]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0.0
  [../]
[]

[Materials]
  [./Copper]
    type = GBEvolution
  
    T = 500 # K
    wGB = 15 # nm
    GBmob0 = 2.5e-6 # m^4/(Js) from Schoenfelder 1997
    Q = 0.23 # Migration energy in eV
    GBenergy = 0.708 # GB energy in J/m^2
  [../]
 
  [./ElasticityTensor]
    type = ComputePolycrystalElasticityTensor
    grain_tracker = grain_tracker
  [../]
  [./strain]
    type = ComputeSmallStrain
    block = 0
    displacements = 'disp_x disp_y'
  [../]
  [./stress]
    type = ComputeLinearElasticStress
    block = 0
  [../]
[]
[Controls]
  [step1]
    type = TimePeriod
    enable_objects = 'Materials::Copper'
    start_time = '0'
    end_time = '5'
  []
  [step2]
    type = TimePeriod
    disable_objects = 'Materials::Copper'
     start_time = '5'
    end_time = '11'
  []
[]  

[Postprocessors]
  [./ngrains]
    type = FeatureFloodCount
    variable = bnds
    threshold = 0.7
  [../]
  [./dofs]
    type = NumDOFs
  [../]
  [./dt]
    type = TimestepSize
  [../]
  [./run_time]
    type = PerfGraphData
    section_name = "Root"
    data_type = total
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    coupled_groups = 'disp_x,disp_y'
  [../]
[]

[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = PJFNK
  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart -pc_hypre_boomeramg_strong_threshold'
  petsc_options_value = 'hypre boomeramg 31 0.7'
  l_tol = 1.0e-4
  l_max_its = 30
  nl_max_its = 25
  nl_rel_tol = 1.0e-7
  start_time = 0.0
  dt=1
  num_steps = 10
  [./Adaptivity]
    initial_adaptivity = 2
    refine_fraction = 0.8
    coarsen_fraction = 0.05
    max_h_level = 3
  [../]
[]

[Outputs]
  exodus = true
[]

```

**Comment by GiudGiud** [comment]
Hello

I can see several options. You could:

Change the equations after the desired time by
- adding a time derivative with a very very high coefficient that prevents any change in the solution 
- use the control system to turn off kernels for the variables of interest. Just keep the time derivative kernel that's it


Create a checkpoint and restart with a different set of equations. 
This is the best option if the variables are different in the second part of the solve. You do not want to have nonlinear variables that you just want to freeze. 

Similarly you can create a multiapp that does the second solve. 
The first solve runs to completion, then executed the multiapp on FINAL to perform the second part. 
You can transfer fields from the first solve to the second one 

Guillaume

**Reply by Yaode-Yin** [reply]
Thanks for your reply, Guillaume! I decide to use the `MultiApp`. But I don't know how to transfer the grain-orientation-related elastic tensor to the subsystem. In the parent system, the elastic tensor is computed by `ComputePolycrystalElasticityTensor`, which functions with the `grain tracker `. But in the subsystem, I stop the grain growth, and want to make a pure mechanical analysis. I think I cannot use `ComputePolycrystalElasticityTensor` in the subsystem. What 's the substitute?  And how can I ensure that each grains retains the same elastic tensor as those in the parent system? 
By the way, I also want to distinguish the grain interiors and boundaries in the subsystem.  I realize `BndsCalc` may be helpful. Can I directly copy the `bnds ` variables from the parent system into the subsystem via `MultiAppCopyTransfer` for distinguishing grain boundaries?

**Reply by GiudGiud** [reply]
Hello

You can transfer variables and the outputs of a class of user objects, the spatial user objects. 
Functions and material properties have to be converted to variables first before being transferred 

If the meshes are the same then yes you can use the MumtiAppCopyTransfer to move the variables. 

Will you be using adaptivity? Keeping the meshes the same with adaptivity requires a little bit of care

Guillaume

**Reply by Yaode-Yin** [reply]
Hello, Guillaume. I plan to use the adaptivity in the subsystem.  _Functions and material properties have to be converted to variables first before being transferred._  Does it mean that I need to transfer c1111 variable? 
```
[AuxVariables]
    [./C1111]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]
[AuxKernels]
 [./C1111]
    type = RankFourAux
    variable = C1111
    rank_four_tensor = elasticity_tensor
    index_l = 0
    index_j = 0
    index_k = 0
    index_i = 0
    execute_on = timestep_end
 [../]
[]
```
This is the a component of the elastic tensor. So I think I need many variables like C1111 and then transfer them into the subsystem. How can I use the transferred variables in subsystem in the mechanical analysis?

**Reply by GiudGiud** [reply]
Hello

Yes you will have to transfer every compoent as variables, then recompose the tensor material property from the variables

I think there is an option for multiapps to use the same mesh as the parent app. This might be what we want for adaptivity. I have never tried
